src/fsharp/opt.fs

```diff
@@ -1300,7 +1298,7 @@ and OpHasEffect g op =
     | TOp.TupleFieldGet(_) -> false
     | TOp.ExnFieldGet(ecref,n) -> isExnFieldMutable ecref n 
     | TOp.RefAddrGet -> false
-    | TOp.ValFieldGet rfref  -> rfref.RecdField.IsMutable
+    | TOp.ValFieldGet rfref  -> rfref.RecdField.IsMutable || HasFSharpAttribute g g.attrib_AllowNullLiteralAttribute rfref.Tycon.Attribs
```

Comment:
Thanks, totally forgot that it could be set as `false`. I'll update and add a test.